File: src/services/instructions/alu/ALU_S_REGX_MREGY_XX.java
package emulator.source.alu

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class ALU_S_REGX_MREGY_XX extends Instruction {
	int type
	
	public ALU_S_REGX_MREGY_XX(short[] memory, int address, int source, int destination, int type) {
		super(memory, address, source, destination)
		super.setArgument32()
		super.setAssembler(Instruction.getTypeStr(type) + this.sdest + ", [" + this.ssource + " + 0x%08x]")
		this.type = type
	}
	
	
	exec ({ context, memory }) {
		int old_a = context.getReg(this.destination) 
		long res = 0
		switch (type) {
		case ADD_S: res = context.getReg(this.destination)  + context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case SUB_S: res = context.getReg(this.destination)  - context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case AND_S: res = context.getReg(this.destination)  & context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case OR_S : res = context.getReg(this.destination)  | context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case XOR_S: res = context.getReg(this.destination)  ^ context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case SHL_S: res = context.getReg(this.destination)  << context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case SHR_S: res = context.getReg(this.destination)  >>> context.memory[fix(context.getReg(this.source)  + this.argument) / 2] break
		case MUL_S:	res = context.getReg(this.destination)  * context.memory[fix(context.getReg(this.source)  + this.argument) / 2] 
					context.h  = (int)((res & 0xffffffff00000000L) >> 32)
					break
		case DIV_S: 	res = context.getReg(this.destination)  / context.memory[fix(context.getReg(this.source)  + this.argument) / 2] 
					context.h  = context.getReg(this.destination)  % context.memory[fix(context.getReg(this.source)  + this.argument) / 2]
					break
		default: throw new RuntimeException("Unsupported operation type: " + type)
		}
		context.getReg(this.destination)  = (int)res
		markFlags(res, context.getReg(this.destination) , context)
		markOverflow(old_a, context.memory[fix(context.getReg(this.source)  + this.argument) / 2], context.getReg(this.destination) , context)
		context.pc  += 6
	}
}

File: src/services/instructions/alu/AluSRegXMRegYXX.js
import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'
import { INSTRUCTIONS_TYPES_NAMES, INSTRUCTION_TYPES } from '@/constants/instructions'
import { INSTRUCTIONS_TYPE_FUNCTION_COMPUTATIONS } from '@/helpers/instruction'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (type, sdestination, ssource) => `${type} ${sdestination}, [${ssource} + %s]`

export default class AluSRegXMRegYXX extends Instruction {
	constructor (memory, address, source, destination, type, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		super.setArgument32(memory)
		super.setAssembler(
			ASSEMBLER_INSTRUCTION_EXPRESSION(
				INSTRUCTIONS_TYPES_NAMES[type],
				this.sdestination,
				this.ssource
			),
			symbolTable
		)

		this.type = type
	}

	exec ({ context, memory }) {
		const old_a = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]
		let result = 0

		result = INSTRUCTIONS_TYPE_FUNCTION_COMPUTATIONS[this.type](
			context[REGISTER_VALUE_NAME_MAPPER[this.destination]],
			memory[Math.floor(Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.source]]  + this.argument) / 2)]
		)

		switch (this.type) {
			case INSTRUCTION_TYPES.MUL_S:
				//TODO: int cast and L number -->> context.h  = (int)((res & 0xffffffff00000000L) >> 32)
				context.h  = (result & 0xffffffff00000000) >> 32
				break
			case INSTRUCTION_TYPES.DIV_S:
				context.h  =
				  context[REGISTER_VALUE_NAME_MAPPER[this.destination]] %
				  memory[Math.floor(Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.source]]  + this.argument) / 2)]
				break
		}

		context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  = result

		Instruction.markFlags(result, context[REGISTER_VALUE_NAME_MAPPER[this.destination]] , context)
		Instruction.markOverflow(
			old_a,
			memory[Math.floor(Instruction.fix(context[REGISTER_VALUE_NAME_MAPPER[this.source]]  + this.argument) / 2)],
			context[REGISTER_VALUE_NAME_MAPPER[this.destination]] , context
		)

		context.pc  += 6
	}
}
